lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details
+
+r"""Interface library for the Prometheus Pushgateway.
+
+This library wraps a relation endpoint using the `pushgwateway` interface
+and exposes an API for forwarding metrics to Prometheus.
+
+
+## Getting Started
+
+To get started using the library, you just need to fetch the library using `charmcraft`.
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.prometheus_pushgateway_k8s.v0.pushgateway
+```
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+requires:
+  your-relation-name:
+    interface: pushgateway
+```
+
+In the source of your charm, first import the interface:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+````
+
+Instantiate the object in your charm's `__init__`, like so:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+from ops.charm import CharmBase
+
+class MyCharm(CharmBase):
+    def __init__(...):
+         ...
+        self.pushgateway_requirer = PrometheusPushgatewayRequirer(self, "your-relation-name")
+```
+
+Then use it at any moment to send a metric (validating that the requirer is ready), passing
+its name and value:
+
+```
+    if self.pushgateway_requirer.is_ready():
+        self.pushgateway_requirer.send_metric("test_metric", 3.141592)
+```
+
+The requirer is ready when the relation to the Prometheus Pushgateway is properly established.
+
+The `send_metric` call will just end quietly if the metric was sent succesfully, or will raise
+an exception if something is wrong (that error should be logged or informed to the operator).
+"""
+
+import json
+import logging
+import socket
+from typing import Optional, Union
+from urllib import request
+from urllib.error import HTTPError
+
+from ops.charm import CharmBase, RelationEvent
+from ops.framework import Object
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# the key in the relation data
+RELATION_KEY = "push-endpoint"
```

Comment:
We've never done it before because of py3.5, but now, should this be encapsulated in a dataclass?

```python
@dataclass
class RelationData:
    push_endpoint: str
```
```python
relation.data[...].update(pushgateway_relation_data)
```

Reply:
I don't think ONE constant is enough to have a dataclass, that would be adding complexity without really gaining anything...